tensorflow_quantum/core/ops/tfq_calculate_unitary_op.cc

```diff
@@ -0,0 +1,163 @@
+    std::vector<SymbolMap> maps;
+    OP_REQUIRES_OK(context, GetSymbolMaps(context, &maps));
+
+    OP_REQUIRES(
+        context, maps.size() == programs.size(),
+        tensorflow::errors::InvalidArgument(absl::StrCat(
+            "Number of circuits and values do not match. Got ", programs.size(),
+            " circuits and ", maps.size(), " values.")));
+
+    int max_num_qubits = 0;
+    for (const int num : num_qubits) {
+      max_num_qubits = std::max(max_num_qubits, num);
+    }
+
+    // TODO(pmassey): Investigate creating a matrix that isn't just the maximum
+    // required size.
+    const int output_dim_size = maps.size();
+    tensorflow::TensorShape output_shape;
+    output_shape.AddDim(output_dim_size);
+    output_shape.AddDim(1 << max_num_qubits);
+    output_shape.AddDim(1 << max_num_qubits);
+
+    tensorflow::Tensor *output = nullptr;
+    OP_REQUIRES_OK(context, context->allocate_output(0, output_shape, &output));
+    auto output_tensor = output->tensor<std::complex<float>, 3>();
+
+    auto DoWork = [&](int start, int end) {
+      std::unique_ptr<UnitarySpace> state = GetUnitarySpace(1, 1);
+      int old_num_qubits = -1;
+      for (int i = start; i < end; i++) {
+        Program program = programs[i];
+        const int num = num_qubits[i];
+        OP_REQUIRES_OK(context, ResolveSymbols(maps[i], &program));
+
+        // QSim work below
+        Circuit circuit;
+        OP_REQUIRES_OK(context, CircuitFromProgram(program, num, &circuit));
+
+        // TODO(mbbrough): Update this allocation hack so that a StateSpace
+        //  object can grow it's memory dynamically to larger and larger size
+        //  without ever having to call free (until the very end). This is
+        //  tricky to implement because right now certain statespaces can't
+        //  simulate all states and we use StateSpaceSlow for smaller circuits.
+        if (num != old_num_qubits) {
+          state = GetUnitarySpace(num, 1);
+          state->CreateUnitary();
+        }
+        state->SetIdentity();
+        OP_REQUIRES_OK(context, state->Update(circuit));
+        uint64_t state_size = uint64_t(1) << state->GetNumQubits();
+        for (uint64_t j = 0; j < state_size; j++) {
+          for (uint64_t k = 0; k < state_size; k++) {
+            output_tensor(i, j, k) = state->GetEntry(j, k);
+          }
+        }
+        // -2 padding for lower portion.
+        for (uint64_t j = state_size; j < (uint64_t(1) << max_num_qubits);
+             j++) {
+          for (uint64_t k = 0; k < (uint64_t(1) << max_num_qubits); k++) {
+            output_tensor(i, j, k) = std::complex<float>(-2, 0);
+          }
+        }
+        // -2 padding for right portion.
+        for (uint64_t j = 0; j < state_size; j++) {
+          for (uint64_t k = state_size; k < (uint64_t(1) << max_num_qubits);
+               k++) {
+            output_tensor(i, j, k) = std::complex<float>(-2, 0);
+          }
+        }
```

Comment:
I hope one of these can remove these padding for-loops by initializing the output tensor with -2:

1. Ln 78, I want to know if it works or not
```cc
    auto output_tensor = output->tensor<std::complex<float>, 3>(std::complex<float>(-2, 0));
```

2. TF provides a functor to fill the given tensor with scalar value
```cc
    const Tensor& Tvalue = [create a tensor with value -2 in the type `std::complex<float>`];
    tensorflow::functor::FillFunctor<Device, std::complex<float>> f;
    f(context->eigen_device<Device>(), out->flat<std::complex<float>>(), Tvalue.scalar<std::complex<float>>());
```

Reply:
Ahhh good point. When we do
```c++
auto output_tensor = output->tensor<std::complex<float>, 3>();
```
We do not loop over the memory to intialize it (which is good because it's faster).
The loops that we use here only iterate over the full tensor once and touch each entry of the tensor exactly once, giving runtime pretty close to `O(2**max_qubits * 2**max_qubits)`
Both of these methods look like the runtime might be
`O(2**n * 2**n) + O(2 ** n * 2 ** n)`
because we overwrite certain memory spaces twice. Although initializing memory to -2 using something fast like `memcpy` or `memset` might be faster. What do you think ?